### ISSUE TITLE
`plot_fit` with matplotlib backend should show model on top of data not below

### ISSUE THREAD

**Issue by hamogu** [sherpa/sherpa#662]
Plotting a dataset with lots of data point with `plot_fit` makes it very hard to see the model, because for some reasons the datapoints are "on top" and the model is "below". 
![image](https://user-images.githubusercontent.com/498688/63176272-a2d09100-c013-11e9-978c-1a50901b1587.png) 
Notice how orange model is essentially invisible. I checked the zorder and tried to go through the code and I don't understand why that happens, but the plotting code is also hard to follow for me. It looks as if the model get plotted after the data, but I'm not sure because there are several steps to follow through from the backend agnostic code to the actual pylab plotting commands.

**Comment by DougBurke** [comment]
I would expect the model plot to be drawn on top. However, it is drawn "per dataset", so

```
  plot_fit("foo")
  plot_fit("bar", overplot=True)
  plot_fit("bax", overplot=True)
```

should (I believe) draw, in order

  - data for foo
  - model for foo
  - data for bar
  - model for bar
  - data for bax
  - model for bax

It isn't clear to me if you want all data drawn and then all models (so that 'data for bax' does not lie on top of 'model for foo'). I can imagine this would be useful (and may require a "plot n datasets" function call, as I kind-of suggested somewhere in https://github.com/sherpa/sherpa/issues/663#issuecomment-524043316 )

The plotting is controlled by the `FitPlot` class - https://sherpa.readthedocs.io/en/latest/plots/api/sherpa.plot.FitPlot.html - and it's "plot" method is

```
    def plot(self, overplot=False, clearwindow=True):
        if self.dataplot is None or self.modelplot is None:
            raise PlotErr("nodataormodel")
        self.dataplot.plot(overplot, clearwindow)
        self.modelplot.overplot()
```

So it draws the data plot and then overplots the model. The `sherpa.plot.DataPlot.plot` method should be

```
    def plot(self, overplot=False, clearwindow=True):
        Plot.plot(self, self.x, self.y, self.yerr, self.xerr, self.title,
                  self.xlabel, self.ylabel, overplot, clearwindow)
```

and the `sherpa.plot.ModelPlot.overplot` method is a wrapper around it's plot method:

```
    def overplot(self, *args, **kwargs):
        "Add the data to an existing plot."
        kwargs['overplot'] = True
        self.plot(*args, **kwargs)

   def plot(self, overplot=False, clearwindow=True):
        Plot.plot(self, self.x, self.y, title=self.title, xlabel=self.xlabel,
                  ylabel=self.ylabel, overplot=overplot,
                  clearwindow=clearwindow)
```

So the actual plot is from `sherpa.plots.Plot.plot`, where we can see the call to the backend plot function:

```
    def plot(self, x, y, yerr=None, xerr=None, title=None, xlabel=None,
             ylabel=None, overplot=False, clearwindow=True):
        backend.plot(x, y, yerr, xerr, title, xlabel, ylabel, overplot,
                     clearwindow, **self.plot_prefs)

```

So, it really should name these arguments rather than using positional order (as I've seen bugs from this), but I think the overplot setting is getting passed over correctly. The `pylab_backend.plot` method starts like

```
def plot(x, y, yerr=None, xerr=None, title=None, xlabel=None, ylabel=None,
         overplot=False, clearwindow=True,
         xerrorbars=False,
         yerrorbars=False,
         ecolor=_errorbar_defaults['ecolor'],
         capsize=_errorbar_defaults['capsize'],
         barsabove=_errorbar_defaults['barsabove'],
         xlog=False,
         ylog=False,
         linestyle='steps',
         linecolor=None,
         color=None,
         marker='None',
         markerfacecolor=None,
         markersize=None,
         xaxis=False,
         ratioline=False):

    if overplot:
        axes = pylab.gca()
    else:

```

I am 99% sure we are using this `plot` method for both data and model, but there's a possibility that the model is drawn as a "histogram" rather than plot. I haven't dug into the pylab plot code to see if it is doing something "strange" that would result in the model being lost.

**Comment by DougBurke** [comment]
The following simple plot

```
from sherpa import ui
ui.load_arrays(1, [10, 20, 30, 40], [5, 6, 5, 4.9])
ui.set_source(ui.const1d.bob)
bob.c0 = 5
ui.plot_fit()
```

creates slightly-different plots with ChIPS and Matplotlib

a) ChIPS

![fit-chips](https://user-images.githubusercontent.com/224638/63610906-98307180-c5a8-11e9-9e28-f03015cebfeb.png)

b) Matplotlib

![fit-mpl](https://user-images.githubusercontent.com/224638/63610914-9e265280-c5a8-11e9-9b26-c6ae0d5d0f68.png)

In ChIPS the model is drawn above the data, but for matplotlib it is above the error bar but behind the points.

**Comment by DougBurke** [comment]
So, in matplotlib when plotting something with errors we do something like

```
ax = plt.gca()
ax.plot(x, y)
ax.errorbar(x, y, grab-color-from-the-first-plot)
```

so in the `plot_fit` example, we have

```
[l.get_zorder() for plt.gca().lines] === [2, 2.1, 2]
```

where the first two values are from `plot` and then `errorbar`, and the last value is the model "curve".

So, the points from the error bars are being drawn at a higher z-order than the lines, which presumably creates the issue @hamogu has reported.

a) do we really need both `ax.plot` and `ax.errorbar`? This may be a hold over from "old" matplotlib behavior

b) how best to handle this? I couldn't find obvious documentation on the matplotlib docs, but it seems that `errorbar` is "special" in that it creates both `Line2D` and `LineCollection`objects, and there's fun in getting the ordering to play nicely together (e.g. see this old bug report https://github.com/matplotlib/matplotlib/issues/1622 which has a discussion on the ordering)

**Comment by DougBurke** [comment]
@hamogu - can you try out the PR in #667 to see if it helps, makes things worse, or breaks something else (these are not mutually-exclusive categories here)?

**Comment by DougBurke** [comment]
This has been addressed in #667.
